### PR TITLE
Minimal period for continuous systems

### PR DESCRIPTION
**Pull request by JonasKoziorek**
I reuse the Poincare map `current_crossing_time` to compute the minimal period of a PO. I create the hyperplane by stepping shortly along the trajectory and using the difference as a normal vector.

### CODE REVIEW

**Review by Datseris — COMMENTED** [review]
Let's add more tests here. Let's add a test for a dynamical system we know analytically its period. For example, in the normal form of the hopf bifurcation, there is this dynamical system dz/dt: https://en.wikipedia.org/wiki/Hopf_bifurcation . From that the period is known analytically, it is 2*π*ω , so we can use this a test. (this is a stable limit cycle).

Let's also move the minimal period at its own dedicated file.

In our last call, have we decided that minimal period should always be called by periodic orbit returning algorithms? Or not, and we leave it to the user to call it?

File: src/api.jl [hunk 1]
```diff
@@ -221,10 +233,23 @@ function _minimal_period(ds::DiscreteTimeDynamicalSystem, po::PeriodicOrbit, ato
     return po
 end
 
-function _minimal_period(ds::ContinuousTimeDynamicalSystem, po::PeriodicOrbit, atol)
-    # TODO
-    minT_po = po
-    return minT_po
+function _minimal_period(ds::ContinuousTimeDynamicalSystem, po::PeriodicOrbit, atol; maxiter=10, stepsize=1e-4)
```
Comment (Datseris) [inline src/api.jl:236]: why is `atol` an argument and the rest are keyword arguments? I think all three should be keywords.

**Comment by JonasKoziorek** [comment]
> In our last call, have we decided that minimal period should always be called by periodic orbit returning algorithms? Or not, and we leave it to the user to call it?

We have decided to always complete the orbit on creation and also always check the local stability. Doing minimal period check shouldn't introduce too much overhead.

I think that minimal period should be called by default on creation. We can add a keyword argument to enable the user to disable this (and the local stability check).

**Comment by JonasKoziorek** [comment]
In a different PR, I will allow the user to enable/disable minimal period check and stability check.

**Comment by JonasKoziorek** [comment]
I am thinking whether it is useful to have function `minimal_period -> minT_PO` that returns a new PO with the minimal period or whether it would be enough to have `minimal_period -> minT` that just returns the minimal period (right now there are both but the second is not exported). If the period is minimal by default in the constructor, the user will not have the need to ever call the first function. What do you think is the best for the user?

**Comment by Datseris** [comment]
I think `minimal_period -> minT_PO` makes more sense. Since the struct is immutable and it references the inner vectors, constructing a new version is practically zero cost. In any case, this doesn't matter so much, since we agreed that this function is called internally, so it is not much use for the user to call it. However, it is important to allow to disable the call for performance (which we can do in a different PR).

In a different PR we can make a high level function `periodic_orbit` and separate a low level function `_periodic_orbit`. The latter just finds the orbit. While, `periodic_orbit` just calls `periodic_orbit` and then calls both the stability function and the minimal period function. This allows us to detach the internal implementations from the estimation of stability and minimal period. We can do something like:


```
function _periodic_orbit

return PO, stability, minimalperiod
```
for the internal function. By default `stability, minimalperiod` can be both `missing`. The high level function `periodic_orbit` then checks and computes the stability and minimal period. This allows the internal algorithms to communitcate the stability of the orbit, if it is known/found while finding the orbit, but also not communicate it if it isn't.

**Comment by Datseris** [comment]
oh, by the way, I think in most literature what we call "minimal period" is called "principal period". I like minimal more, but we should just put a quick note in the docstring of `minimal_period` about this.

**Comment by JonasKoziorek** [comment]
> oh, by the way, I think in most literature what we call "minimal period" is called "principal period". I like minimal more, but we should just put a quick note in the docstring of minimal_period about this.

I mentioned prime, principal and fundamental period.

File: src/minimal_period.jl [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    minimal_period(ds::DynamicalSystem, po::PeriodicOrbit; kw...) → minT_po
+
+Compute the minimal period of the periodic orbit `po` of the dynamical system `ds`.
+Return the periodic orbit `minT_po` with the minimal period. In the literature, minimal 
+period is also called prime, principal or fundamental period.
+
+## Keyword arguments
+
+* `atol = 1e-4` : After stepping the point `u0` for a time `T`, it must return to `atol` neighborhood of itself to be considered periodic.
+* `maxiter = 10` : Maximum number of Poincare map iterations. Continuous-time systems only.
```
Comment (Datseris) [inline src/minimal_period.jl:15]: What do you do if you exceed this? What is returned as the period? Should it error? It is not clear from the docs what happens.

I think this number should be much higher than 10.

Comment (JonasKoziorek) [reply src/minimal_period.jl]: Right now I return the original period `po.T`. I think warning is better than error. I will update `maxiter` to 40.

**Review by Datseris — APPROVED** [review]

**Comment by Datseris** [comment]
fantastic

**Pull request merged**